### PR TITLE
Updates debian dependencies.

### PR DESCRIPTION
**Pull request by traud**
required for Debian 10, Ubuntu 18.10, Ubuntu 19.04, and Ubuntu 19.10 already

### CODE REVIEW

**Comment by damencho** [comment]
You know that this will not solve the problem that you will not be able to load ffmpeg there, right?
This will only solve the problem with installation, but media will not work.

**Comment by traud** [comment]
No, because ‘everything’ worked for me. Which media is affected exactly?

**Comment by damencho** [comment]
Video calls? Do you say that the binaries which are linked to libavcodec57 work fine with libavcodec58? And everything is resolved and working. 
What is the output of ldd libjnffmpeg.so, should be somewhere in /usr/share/jitsi/ ....

**Comment by traud** [comment]
Where do I get that library? I do not have it, when I do a clean install. I do get it in ~/.jitsi/sip-communicator.bin when I install the latest stable and then the latest nightly (5618). However, even in Ubuntu 18.04 LTS or when I re-download it from [here](https://github.com/jitsi/jitsi/tree/5716bdae1b723675195eed99b995ec84cdea7fd0/lib/native/linux-64), that library just points to the standard C libs.

**Comment by damencho** [comment]
This is the old static linked library. Oh the lib is inside jar file you will not find it, my fault. You can download it from https://github.com/jitsi/jitsi-lgpl-dependencies/tree/master/src/main/resources/linux-x86-64

**Comment by damencho** [comment]
On that link there are two libs one linked to older ffmpeg and one to the newer, the versions you see in the control file from master.

**Comment by traud** [comment]
Mhm, now I see. `ldd` gives ‘not found’. I have to dig deeper in all this the next week anyway, because I have an issue with the audio codec AMR-WB. Since the Nightlies, AMR-WB does not work at all, even in Ubuntu 18.04 LTS:
```
Cannot build a flow graph with the customized options:
  Unable to transcode format: LINEAR, 44100.0 Hz, 16-bit, Mono, LittleEndian, Signed
    to: AMR-WB/rtp, 16000.0 Hz, Mono
```
That worked in the latest stable (although with many issues as well).

**Comment by damencho** [comment]
AMR was removed from the latest binaries. 
And to make new ffmpeg binaries that they work for new ffmpeg we need to recompile and add them ... not sure when this will happen.

**Comment by damencho** [comment]
The AMR was disabled in this PR https://github.com/jitsi/libjitsi/pull/403 
https://github.com/jitsi/libjitsi/pull/403/commits/fe0f44aa8a36023c1b69edd8122844608f403b27#diff-8b65e7414388ba981937ebdcdc2dc8e1R111
So if you need it, you will need to add it to ffmpeg and enable the encoder and decoder classes in libjitsi.

**Comment by traud** [comment]
Puh. That was valuable information! Thanks to those starting points, I am now able to use AMR-WB in Ubuntu 19.10 (FFmpeg 4.x) with the latest Jitsi Desktop. Finally, I can look at traud/asterisk-amr/issues/17 and #613.

I love to help, however, I am unsure how to proceed here. Because Java allows dynamic loading of libraries, and because Jitsi Desktop even works without FFmpeg, I wonder if those dependencies ‘matter’ at all. Matter means here, being required. Would it not make more sense to move those dependencies from Depends to [Recommends](https://www.debian.org/doc/manuals/debian-reference/ch02.en.html#_package_dependencies)? In that case, we could proceed with this Pull Request, leaving the other puzzle pieces untouched, yet.

Beside that, when a media codec like AMR-WB is disabled in libjitsi (or not available because FFmpeg was not linked), it has not to show up in the user interface of Jitsi Desktop (or at least being a disabled/grey option with a tool tip explaining the situation). Otherwise users (like me at the start) get no media, if they enabled AMR-WB for example. I guess, something like this—grey media codecs, when not available—is not broken but not implemented, yet.

**Comment by ibauersachs** [comment]
Although I won't have time/motivation to work on it, I still think a lot about these issues. One issue is that we currently have one Debian repo with binaries, only differentiating for stable/nightly builds. This was fine while we provided all native stuff as statically linked binaries. Going forward, IMO we should reconsider how the repo is set up and the build process in general. Now with Github Actions and Docker, there are new options. E.g.
- Build and publish jitsi-ffmpeg for Windows/Mac and a statically linked Linux version as Maven artifacts. For Linux, build the native part as an additional library and publish it for various distros (debian/stretch, debian/buster, ubuntu/xenial, etc., maybe some RPMs)
- Build Jitsi Desktop with a proper build system (Gradle/Maven, CMake) *for each* target, and publish to the repo corresponding to which it was built with.
- Or just pull all (native) dependencies as source and build within Jitsi Desktop's build system. _Maybe_ use git submodules for that. But still build from scratch for every target.

In short, one binary for all Linux just isn't going to work (anymore).

**Comment by traud** [comment]
Puh. I do not understand a word (what is my role, what should I do, why). Actually, I understand so less, I cannot even ask a clever question. Here, we are about the package for Debian and its derivates. Let us start with this:
> one binary for all [Debian derivates] just isn't going to work

Can you elaborate why that does not work?

Guessing around: Are you unhappy with that LGPL.jar? Currently, it has two shared libraries. It needs a third library. Actually building this manually is terrible easy after figuring out all [puzzle pieces…](https://community.jitsi.org/t/18436) would be nice if somebody could comment, if I made everything right. Especially https://github.com/jitsi/libjitsi/commit/204d1f80913d6baadb572e91a39a75399c19bf43 is causing me trouble—and I do not understand how to use Maven within Jitsi Desktop (to avoid that `cp` and grab any locally installed packages instead).

**Comment by ibauersachs** [comment]
>Let us start with this:
>> one binary for all [Debian derivates] just isn't going to work
> Can you elaborate why that does not work?

Because of the different ffmpeg versions in Ubuntu 16.04 / 18.04. For some time, Debian and Ubuntu used the ffmpeg fork libav, then switched back to ffmpeg proper, but exported the symbols differently. 

The the newer the OS release, the newer ffmpeg. And they deprecated quite some APIs we use(d), and the newer ones aren't yet available in older distros. We could argue to drop support for 16.04, but that doesn't solve all issues.

**Comment by traud** [comment]
Mhm, the current solution with two libs works quite well for FFmpeg 2 and FFmpeg 3. It ‘just’ needs another lib for FFmpeg 4. I tested Ubuntu 16.04 LTS, 18.04 LTS and 19.10 in the last few days and was quite happy. Playing around with different build systems, docker or so, does not make the source code more stable, which has to support all three FFmpeg (or drop one or two).

All I faced, was one issue with `jawtrenderer` which did not resolve `libjawt.so`. I had to add the RPath for Java 8 in the `libjitsi/src/native/build.xml`. Not sure, if I did something wrong because that was an issue on all my three Ubuntu versions.

Nevertheless, beside FFmpeg, the same applies to OpenSSL. Only Ubuntu 16.04 LTS works with the current OpenSSL 1.0.x adapter. Since 18.04 LTS, OpenSSL 1.1.x is used. I was able to patch that as well. However, the resulting package does not have that several-libs-trick like FFmpeg has already. Therefore, the package works only with Ubuntu 16.04 LTS _or_ Ubuntu 18.04 LTS (and newer). Not sure if that OpenSSL is used with Jitsi Desktop at all. Well, because of the dynamic loading/binding in Java, this does not matter either. If OpenSSL is found, it is used. If not, Jitsi Desktop starts anyway and falls back to Bouncy Castle.

Consequently, I am bit confused. You _have already_ a solution to support multiple versions of an external shared library. The question is not the build system (or the resulting artifacts) but whether you want to maintain the source code for several versions. And actually, the changes for OpenSSL 1.1 and FFmpeg 4 were slim (two symbols each); at least when it comes to compile time. I have not tested run time.

**Comment by ibauersachs** [comment]
I haven't properly looked at your patch in the -lgpl repo, but it seemed like you simply disabled something. This won't work at runtime.
The solution with multiple binaries we have now were manually compiled and put together. This was a time consuming task, and since it was probably more than a year ago, neither me nor Damencho would be able to do it again. Updating a minor version, accepting a pull request, etc. are all way to tedious now.

And yes, the same problems apply to the OpenSSL stuff. But that's not used in Jitsi Desktop.

**Comment by traud** [comment]
> This won't work at runtime.

Please, double-check if anything uses those two symbols at all. I tested just AMR-WB and that works. I researched and mentioned the originating relevant commits, so you see what the alternatives are. As I do not use video – I do not even have a camera –, I cannot help further on that.

Anyway, changes – coping with the _latest_ version of an external library – cannot be fixed with another build system. This is simply work. You are about the decision whether you want to support _multiple_ versions of external libraries (or drop FFmpeg all together).

This change here is just about getting Jitsi Desktop installed at all. If the optional parts are moved from `Depends` to `Recommends`, this gets even more forward compatible.

**Comment by ibauersachs** [comment]
> > This won't work at runtime.
> 
> Please, double-check if anything uses those two symbols at all. I tested just AMR-WB and that works. I researched and mentioned the originating relevant commits, so you see what the alternatives are. As I do not use video – I do not even have a camera –, I cannot help further on that.

I'm not sure when I find time for that.

> Anyway, changes – coping with the _latest_ version of an external library – cannot be fixed with another build system. This is simply work. You are about the decision whether you want to support _multiple_ versions of external libraries (or drop FFmpeg all together).

Correct, incorporating the latest changes is just work. But making these changes _available_ without too much manual intervention requires a sane build system. Or rather just _a_ build system. Currently we have neither.

> This change here is just about getting Jitsi Desktop installed at all. If the optional parts are moved from `Depends` to `Recommends`, this gets even more forward compatible.

I'm fine with that.

Actually, given all the issues with the video and screensharing support, it might be an option to completely remove that. Jitsi Meet is the better alternative for that.

**Comment by traud** [comment]
> This won’t work at runtime.

In Linux, late binding happens. Only when the symbol is used, a segmentation fault happens. However, the shared library must be able to resolve its depended libraries = files on disk, path in scope.

I rechecked. None of the two symbols are ever called within the Jitsi universe. At least, I did not find anything via a GitHub search. Of course, an external library might use it. Not sure if you have any ‘contracts’ for that library. If yes, there has to be some versioning like a SONAME bump. Probably it is better to remove those two symbols completely. Then, if there is an issue, it reveals more obvious because all platforms are affected.

> making these changes available without too much manual intervention requires a sane build system

OK. I researched that again and bought myself a camera. Here are my findings: Currently, FFmpeg is used for the media format H.264 _or_ AMR-WB. Some cameras come with native H.264 output. That is not used. FFmpeg comes with its own H.264 implementation. That is not used. Instead, OpenH264 is used through FFmpeg. Therefore, a custom **bundled** FFmpeg is used. That bundled FFmpeg does not come with AMR-WB. Therefore it is H.264 _or_ AMR-WB right now. According to my tests and review of the source code, only those two media formats require FFmpeg. By now (see jitsi/libjitsi#498), these two media formats are detected at runtime. Consequently, this is a no-brainer.

However, FFmpeg is also used to enable camera/video support in general. This is an optional feature. Jitsi Desktop starts and can be used without video, just for audio – at least according to my tests. There is only a cosmetic issue on the command line, where the exception about the unsatisfied link is printed in raw, rather than with a human-readable error message. Consequently, it is OK to move FFmpeg from `Depends` to `Recommends`. I changed my Pull Request accordingly.

Now comes the FFmpeg version issue. Currently, the LGPL shared library bridges FFmpeg 2 and FFmpeg 3. This is done through one shared library for each version for each platform. That sounds a lot. However, this affects just Linux, because all _other_ platforms get that bundled FFmpeg (with OpenH264). In Linux, there are currently two platforms: `i386` and `amd64`. Consequently, those are three versions on two platforms, resulting in six shared libraries. Each library can be built with one single command: `ant ffmpeg -DskipOpenh264=true`.

When do you build those libraries:
A) When there is a bug in the wrapper (happened last ? in Dec. 2017). Or
B) When there is a new version of FFmpeg (happened last ? in Apr. 2018)
The last case happens right now: Besides FFmpeg 2 and FFmpeg 3, FFmpeg 4 needs to be added (which offers H.265 and AV1 support, by the way). Because in the mid-term, the bundled FFmpeg must be updated to the latest version, supporting that new version is not adding any additional overhead. That was clarified above already. The problem is how to keep support for the old versions. This is not a problem, because those old wrapper libraries need to be re-build just in case A. In case B, new ones are added, the old ones are not touched.

I hope this explains why I have problems to understand why an _automatic_ build system is needed. I can offer to build those two wrapper libraries for FFmpeg 4. Actually, for `amd64`, I have it already here. It is just creating a virtual machine, install Debian Buster with the right platform†, issue the above command. Then, those files must be added to the JAR: `mvn package`. Then, this third variant must be loaded dynamically, as well. Done.

† I would not even use those cross-compiling features to build `i386`.

**Comment by traud** [comment]
Now, there are several alternatives to get this Pull Request merged:
(A) move the FFmpeg dependencies from `Depends` to `Recommends`,
(B) beside FFmpeg 2.8 and 3.x, add FFmpeg 4.x to the dependencies,
(C) wait until the FFmpeg 4.x wrapper was added with jitsi/jitsi-lgpl-dependencies#10.

Currently, it seems to me, we wait for all three conditions to fulfill. What about going just for alternative A, first? That would allow the installation for those on newer Linux platforms. They are able to run Jitsi Desktop because video is just optional. And still, the dep file would indicate that FFmpeg 4.x is not supported out-of-the-box, yet.

Then, when condition C is fulfilled, I create a new Pull Request here which requests B. What about that? With that approach we do not have so long lived Pull Requests no one benefits from. Furthermore, it would move the discussion to the repositories where it belongs to.

Actually, there is an alternative D:
In Debian, meta packages exist `libavcodec-extra` and `libavfilter-extra`. Those install all other packages (avutil, avformat, …). I am not a fan of that because† it does not indicate that LibJitsi is not working with those. Anyway, I thought I should mention it. On the other hand, those meta packages would avoid to touch this file anytime soon.
† another reason against D: `libavfilter-extra` does not work for FFmpeg 2.8. One approach for those platforms would be not to go for the lib packages but just for the app package `ffmpeg`. That package installs all FFmpeg packages for Jitsi Desktop for sure. Personally, I do not like that either, either.

**Comment by traud** [comment]
OK, I went for alternative A to get this Pull Request done. On Debian 10 and Ubuntu 20.04 LTS, this change allows Jitsi Desktop to be run, although without video (and AMR-WB) support, yet. With this change, an user is able to simply double-click the `.deb` file after installing [jitsi-archive-keyring](https://download.jitsi.org/unstable/jitsi-archive-keyring_1.0.1_all.deb).

By the way, for what feature is that keyring package required? If we get that optional as well, users are able to install Jitsi Desktop even without Int**er**net. I know companies who use Jitsi Desktop 2.10 in their Int**ra**net.

**Comment by damencho** [comment]
> By the way, for what feature is that keyring package required?

This is just adding the public key for the repo and adding the stable repo to `etc/apt/sources.list.d`. This one can be optional, we used it at some point in the past when changing the Debian repository link so we don't break updates for the old version while transitioning to the new repo.

**Comment by traud** [comment]
Does anything block this request? Recommended commit message:

Debian: Depend on FFmpeg and AppIndicator optionally.
- AppIndicator is enabled optionally via Menu » Tools » Options » General » Tray implementation.
- FFmpeg is used for video and the audio codec AMR-WB (therefore, libavcodec-extra preferred).
- jitsi-archive-keyring was used so updates do not break while transitioning to a newer Debian repository.

Jitsi Desktop runs without those. Furthermore, Jitsi changed to Java 8 as minimum, which is reflected here now.

**Comment by Banellica** [comment]
Dear guys

I need to install Jitsi Client on Ubuntu 18.04 (and it complains: "unmet dependencies: jitsi-archive-keyring not installable").

I'd like to just use the audio chat functionality (no need for video).

I am not an expert. Can you provide a simple prescription - what and how should I install?

**Comment by traud** [comment]
> Ubuntu 18.04 LTS complains: "unmet dependencies: jitsi-archive-keyring not installable"

There, [you go …](http://github.com/jitsi/jitsi/issues/540#issuecomment-600524696)
Alternatively, for Ubuntu 18.04 LTS, install just the [latest keyring](https://download.jitsi.org/stable/jitsi-archive-keyring_1.0.1_all.deb) and then [go for this …](http://github.com/jitsi/jitsi/issues/400#issuecomment-565431096)
